pkg/model/mselection/heaviest.go

```diff
@@ -20,137 +19,230 @@ import (
+			}}
 			best.count = c
 		} else if c == best.count {
-			best.tips = append(best.tips, tip.hash)
+			// add the tip to the slice of currently best tips
+			best.tips = append(best.tips, &selectedTip{
+				item:  tip,
+				index: index,
+			})
 		}
 	}
 
-	// TODO: is it really to select a random tip? Maybe prefer the older (or younger) tip instead
-	return randomTip(best.tips), nil
+	selected, err := randomTipFromTips(best.tips)
+	if err != nil {
+		return nil, 0, err
+	}
+	return selected, best.count, nil
 }
 
-// SelectTip selects a tip to be used for the next milestone.
-func (s *HeaviestSelector) SelectTip() (hornet.Hash, error) {
+// SelectTips tries to collect tips that confirm the most transactions in the future cone.
+// best tips are determined by counting the referenced transactions (heaviest branches) and by "removing" the
+// transactions of the referenced cone of the already choosen tips in the bitsets of the available tips.
+// only tips are considered that were present at the beginning of the SelectTips call,
+// to prevent attackers from creating heavier branches while we are searching the best tips.
+// "maxHeaviestBranchTipsPerCheckpoint" is the amount of tips that are collected if
+// the current best tip is not below "UnconfirmedTransactionsThreshold" before.
+// selecting at least one tip can be enforced, even if none of the heaviest branches matches the
+// "minHeaviestBranchUnconfirmedTransactionsThreshold" criteria with "enforceTips".
+// if at least one heaviest branch tip was found, "randomTipsPerCheckpoint" random tips are added
+// to add some additional randomness to prevent parasite chain attacks.
+func (s *HeaviestSelector) SelectTips(enforceTips bool) (hornet.Hashes, error) {
+
+	// copy the tips to release the lock to allow faster iteration
+	// and to get a frozen view of the tangle, so an attacker can't
+	// create heavier branches while we are searching the best tips
+	tipsList := s.copyTipItemsToList()
 
-	if !tangle.IsNodeSynced() {
-		return nil, tangle.ErrNodeNotSynced
+	// tips could be empty after a reset
+	if tipsList.Len() == 0 {
+		return nil, ErrNoTipsAvailable
 	}
 
-	// run the tip selection for at most 0.1s to keep the view on the tangle recent; this should be plenty
-	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(100*time.Millisecond))
-	defer cancel()
-	return s.selectTip(ctx)
+	var result hornet.Hashes
+
+	for i := 0; i < s.maxHeaviestBranchTipsPerCheckpoint; i++ {
+		// run the tip selection for at most 0.1s to keep the view on the tangle recent; this should be plenty
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(100*time.Millisecond))
+
+		tip, count, err := s.selectTip(ctx, tipsList)
+		cancel()
+		if err != nil && err != context.DeadlineExceeded {
+			break
+		}
+
+		// if we want to enforce tips, at least collect one
+		if count < uint(s.minHeaviestBranchUnconfirmedTransactionsThreshold) && !(enforceTips && len(result) == 0) {
```

Comment:
If I understand this condition correctly: when you are enforcing tips and you have at least one result -> break out of the heviest tipsel loop, because you may never / soon enough find a tip that meets the `minHeaviestUnconfirmedTransactionsThreshold`.
For me your condition reads: "if I don't meet the threshold, and either I am not enforcing tips or I have at least a result then break".